### DISCUSSION TITLE
Statusbar widget

### DISCUSSION THREAD

**Discussion by tristan957** [ghostty-org/ghostty#2421] [Feature Requests, Ideas]
Mitchell added a pending key UI for macOS in https://github.com/ghostty-org/ghostty/pull/2418. I think this widget would be a good candidate for a statusbar element.

- Are others interested in such a thing?
- What are other widgets we could add?
- What would the configuration syntax look like?

Widgets that have been suggested:
- Date/Time
- Terminal size
- Cursor position
- Modifier state
- Pending key
- CPU usage
- Memory usage

`toggle_statusbar` action so it can be hidden or displayed when needed.

**Comment by tristan957** [comment]
```
status-bar = on/off
# | splits left, right, and middle, > separates widgets within a group 
status-bar-widgets = pending_key>widget_1>widget_2|widget_3|widget_4>widget_5>widget_6
status-bar-location = top/bottom
```

**Reply by tristan957** [reply]
Statusbars having a left, center, right is a pretty standard approach. See vim/neovim for prior art.

**Reply by mitchellh** [reply]
At least on macOS, it'd be trivial to support spacers rather than explicit L/M/R. So you could, for example, have four separate areas if you wanted. Using your syntax, `|` could also mean spacer. Bikeshedding that, I think its more common and descriptive  to use something like `<->`.

I think an additional question: what do the status bars do when there isn't enough window width to render the widgets? Do they cut them off on the right?

**Reply by tristan957** [reply]
Great question. Here is what Builder does:


https://github.com/user-attachments/assets/43e20177-d0dc-4a42-999d-bb2c3d4ad3ab

**Reply by tristan957** [reply]
How should we configure the date/time format?

```
statusbar-date-time-format = xyz
```
or something inline?

**Reply by tristan957** [reply]
For the statusbar widget syntax we could do something like:

```
statusbar = widget_1
statusbar = widget_2
statusbar = spacer
statusbar = widget_3
...
```
This would match the behavior of keybind and other ghostty config thigs.

**Reply by jcollie** [reply]
How about taking advantage of multiple lines:

```
statusbar-widget = time:left:HH:mm
statusbar-widget = cursor_pos:right
statusbar-widget = pending_keys:center
statusbar-widget = screen_size:right
```
Also, _no_ Go-style time formatting.

**Reply by tristan957** [reply]
> Also, no Go-style time formatting.

strftime formatting for sure

**Reply by tristan957** [reply]
I think the style you suggest, :right, :left, :center would not fit what Mitchell said above. But I do like what you are doing with the config of the date time.

**Reply by mitchellh** [reply]
I think if we separate out _naming/configuring_ widgets from _layout_ then it can all work together. A lot of things I've used in the past with widgets will use their type as a default name and I think that's a good pattern. So simply following along above, I could imagine a syntax like this:

```
status-bar-widget = time:format=HH:MM
status-bar-widget = cursor_pos:name="pos"
status-bar = time<->pos
```

Again, not married at all to the syntax. But I think semantically separating config from layout is going to be cleanest.

**Comment by jcollie** [comment]
Possible widgets:

1. terminal size (rows x cols)
2. cursor pos
3. modifier state (alt, ctrl, shift, caps)
4. date/time

**Comment by PeterCardenas** [comment]
I use tmux to show CPU usage, RAM usage and current time and date

**Comment by jcollie** [comment]
There should be an action that can toggle the visibility of the status bar with the default for new windows being controlled by the config.

**Comment by mitchellh** [comment]
For future note when we support a plugin system, "status bar widget" is also a great place for plugins to expose themselves.

**Comment by oleastre** [comment]
I would like to suggest adding a "string" kind of widget whose content could be set through an action. 
In my terminal config, I use tmux to display a bottom status-bar with current path(in full), current git branch, and  some other info like python environments.
All of that is set in a PROMPT_COMMAND which calls `tmux set-option -p pane-border-format " $tmux_line ";`
I could imagine a similar `ghostty` call to set the status bar string

**Reply by tristan957** [reply]
There will be a scripting API in the future that can tackle this

**Comment by XhstormR** [comment]
Currently, I'm using a custom tmux status line to monitor memory usage and system uptime, host ip, network status, etc.

<img width="1412" height="530" alt="image" src="https://github.com/user-attachments/assets/ddf8d47d-7f70-4d3d-a3ec-1f003cd0468f" />

**Comment by kendfss** [comment]
I, for one, would love to have widgets for tabs and battery level, assuming this isn't going to be deferred to the plugin system.
